File: desktop/WaiComputer.Core/Dictation/DictationDictionaryStore.cs
using System.Linq;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using WaiComputer.Core.Api;
using WaiComputer.Core.Api.Models;
using WaiComputer.Core.Time;

namespace WaiComputer.Core.Dictation;

/// <summary>
/// A personal-dictionary entry. With no replacement it is a vocabulary booster
/// (improves recognition); with a replacement it is an auto-correction applied
/// after transcription. Ports the macOS <c>DictionaryWord</c>.
/// </summary>
public sealed record DictionaryWord(Guid Id, string Word, string? Replacement, DateTimeOffset CreatedAt)
{
    public bool IsReplacement => Replacement is not null && Replacement != Word;
}

/// <summary>
/// Local-first dictation dictionary: vocabulary hints for the provider +
/// post-transcription replacements. Cached via <see cref="IDictationLocalStore"/>
/// (survives logout) and best-effort synced to the backend. Ports the macOS
/// <c>DictationDictionaryStore</c>. Remote failures are logged and retried on the
/// next <see cref="HydrateAsync"/> — the local cache is the session source of truth.
/// </summary>
public sealed class DictationDictionaryStore
{
    private const string WordsKey = "dictation_dictionary";
    private const string TombstonesKey = "dictation_dictionary_tombstones";

    private readonly IDictationLocalStore _local;
    private readonly IApiClient _api;
    private readonly ISystemClock _clock;
    private readonly ILogger<DictationDictionaryStore> _logger;
    private readonly object _gate = new();
    private List<DictionaryWord> _words = new();
    private HashSet<Guid> _tombstones = new();

    public DictationDictionaryStore(IDictationLocalStore local, IApiClient api, ISystemClock clock, ILogger<DictationDictionaryStore>? logger = null)
    {
        _local = local;
        _api = api;
        _clock = clock;
        _logger = logger ?? NullLogger<DictationDictionaryStore>.Instance;
    }

    public IReadOnlyList<DictionaryWord> Words
    {
        get { lock (_gate) { return _words.ToList(); } }
    }

    /// <summary>Vocabulary list sent to the transcription provider for prompt conditioning.</summary>
    public IReadOnlyList<string> VocabularyList
    {
        get { lock (_gate) { return _words.Select(w => w.Word).ToList(); } }
    }

    public async Task LoadAsync(CancellationToken ct)
    {
        var words = await _local.ReadAsync<List<DictionaryWord>>(WordsKey, ct).ConfigureAwait(false) ?? new List<DictionaryWord>();
        var tombs = await _local.ReadAsync<List<Guid>>(TombstonesKey, ct).ConfigureAwait(false) ?? new List<Guid>();
        lock (_gate) { _words = words; _tombstones = new HashSet<Guid>(tombs); }
    }

    /// <summary>Apply case-insensitive replacement rules to transcribed text (vocabulary boosters are not substituted).</summary>
    public string ApplyReplacements(string text)
    {
        List<DictionaryWord> words;
        lock (_gate) { words = _words.Where(w => w.IsReplacement).ToList(); }
        var result = text;
        foreach (var word in words)
        {
            result = result.Replace(word.Word, word.Replacement!, StringComparison.OrdinalIgnoreCase);
        }
        return result;
    }

    public async Task AddAsync(string word, string? replacement, CancellationToken ct)
    {
        var trimmed = word.Trim();
        if (trimmed.Length == 0)
        {
            return;
        }

        DictionaryWord entry;
        lock (_gate)
        {
            if (_words.Any(w => string.Equals(w.Word, trimmed, StringComparison.OrdinalIgnoreCase)))
            {
                return; // dedup by word (case-insensitive)
            }
            entry = new DictionaryWord(Guid.NewGuid(), trimmed, replacement, _clock.UtcNow);
            _words.Add(entry);
            _words.Sort((a, b) => string.Compare(a.Word, b.Word, StringComparison.OrdinalIgnoreCase));
        }

        await SaveWordsAsync(ct).ConfigureAwait(false);
        await PushWordAsync(entry, ct).ConfigureAwait(false);
    }

    public async Task DeleteAsync(Guid id, CancellationToken ct)
    {
        lock (_gate)
        {
            _words.RemoveAll(w => w.Id == id);
            _tombstones.Add(id);
        }
        await SaveWordsAsync(ct).ConfigureAwait(false);
        await SaveTombstonesAsync(ct).ConfigureAwait(false);
        try
        {
            await _api.DeleteDictionaryWordAsync(id, ct).ConfigureAwait(false);
            lock (_gate) { _tombstones.Remove(id); }
            await SaveTombstonesAsync(ct).ConfigureAwait(false);
        }
        catch (Exception ex) { _logger.LogWarning(ex, "Delete dictionary word failed; tombstone retained"); }
    }

    /// <summary>Pull server state, replay tombstoned deletes, merge server-only words, push local-only words.</summary>
    public async Task HydrateAsync(CancellationToken ct)
    {
        IReadOnlyList<DictionaryWordDto> server;
        try { server = await _api.ListDictionaryAsync(ct).ConfigureAwait(false); }
        catch (Exception ex) { _logger.LogWarning(ex, "Hydrate dictionary fetch failed"); return; }

        HashSet<Guid> tombs;
        HashSet<Guid> localIds;
        lock (_gate) { tombs = new HashSet<Guid>(_tombstones); localIds = _words.Select(w => w.Id).ToHashSet(); }

        foreach (var s in server.Where(s => tombs.Contains(s.ClientWordId)))
        {
            try { await _api.DeleteDictionaryWordAsync(s.ClientWordId, ct).ConfigureAwait(false); lock (_gate) { _tombstones.Remove(s.ClientWordId); } }
            catch (Exception ex) { _logger.LogWarning(ex, "Hydrate tombstone replay failed"); }
        }
        await SaveTombstonesAsync(ct).ConfigureAwait(false);

        var additions = server
            .Where(s => !localIds.Contains(s.ClientWordId) && !tombs.Contains(s.ClientWordId))
            .Select(s => new DictionaryWord(s.ClientWordId, s.Word, s.Replacement, s.OccurredAt))
            .ToList();
        if (additions.Count > 0)
        {
            lock (_gate)
            {
                _words.AddRange(additions);
                _words.Sort((a, b) => string.Compare(a.Word, b.Word, StringComparison.OrdinalIgnoreCase));
            }
            await SaveWordsAsync(ct).ConfigureAwait(false);
        }

        var serverIds = server.Select(s => s.ClientWordId).ToHashSet();
        List<DictionaryWord> localOnly;
        lock (_gate) { localOnly = _words.Where(w => !serverIds.Contains(w.Id)).ToList(); }
        foreach (var entry in localOnly)
        {
            await PushWordAsync(entry, ct).ConfigureAwait(false);
        }
    }

    public async Task ClearLocalCacheAsync(CancellationToken ct)
    {
        lock (_gate) { _words.Clear(); _tombstones.Clear(); }
        await SaveWordsAsync(ct).ConfigureAwait(false);
        await SaveTombstonesAsync(ct).ConfigureAwait(false);
    }

    private async Task PushWordAsync(DictionaryWord entry, CancellationToken ct)
    {
        try { await _api.CreateDictionaryWordAsync(new CreateDictionaryWordRequest(entry.Id, entry.Word, entry.Replacement, entry.CreatedAt), ct).ConfigureAwait(false); }
        catch (Exception ex) { _logger.LogWarning(ex, "Push dictionary word failed; will retry on next hydrate"); }
    }

    private Task SaveWordsAsync(CancellationToken ct)
    {
        List<DictionaryWord> snapshot;
        lock (_gate) { snapshot = _words.ToList(); }
        return _local.WriteAsync(WordsKey, snapshot, ct);
    }

    private Task SaveTombstonesAsync(CancellationToken ct)
    {
        List<Guid> snapshot;
        lock (_gate) { snapshot = _tombstones.ToList(); }
        return _local.WriteAsync(TombstonesKey, snapshot, ct);
    }
}

File: ios/WaiComputer/WaiComputer/Features/Dictation/DictationDictionaryStore.swift
import Foundation
import os
import WaiComputerKit

private let log = Logger(subsystem: "is.waiwai.computer.app", category: "dictation-dictionary")

struct DictionaryWord: Identifiable, Codable, Hashable {
    let id: UUID
    var word: String
    var replacement: String?
    let createdAt: Date

    init(word: String, replacement: String? = nil) {
        self.id = UUID()
        self.word = word
        self.replacement = replacement
        self.createdAt = Date()
    }

    init(id: UUID, word: String, replacement: String?, createdAt: Date) {
        self.id = id
        self.word = word
        self.replacement = replacement
        self.createdAt = createdAt
    }

    /// Words without replacement are vocabulary boosters (improve recognition).
    /// Words with replacement are auto-corrections applied after transcription.
    var isReplacement: Bool {
        replacement != nil && replacement != word
    }
}

/// Server-synced custom dictation dictionary. Ported verbatim from macOS (pure
/// Foundation/os/WaiComputerKit). Mirrors the server via the shared
/// `listDictationDictionary` / `createDictionaryWord` / `deleteDictionaryWord`
/// endpoints. The view layer is iOS-native; this store is identical.
@MainActor
final class DictationDictionaryStore: ObservableObject {
    @Published private(set) var words: [DictionaryWord] = []

    private let fileURL: URL
    private let tombstonesURL: URL
    private var tombstones: Set<UUID> = []
    private var apiClient: APIClient?

    init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = appSupport.appendingPathComponent("WaiComputer", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent("dictation_dictionary.json")
        self.tombstonesURL = dir.appendingPathComponent("dictation_dictionary_tombstones.json")
        load()
        loadTombstones()
    }

    // MARK: - Sync wiring

    func attach(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func add(word: String, replacement: String? = nil) {
        let trimmed = word.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        guard !words.contains(where: { $0.word.lowercased() == trimmed.lowercased() }) else { return }

        let entry = DictionaryWord(word: trimmed, replacement: replacement)
        words.append(entry)
        words.sort { $0.word.localizedCaseInsensitiveCompare($1.word) == .orderedAscending }
        save()
        log.info("Added dictionary word: \(trimmed)")

        if apiClient != nil {
            Task { await self.pushWord(entry) }
        }
    }

    func delete(_ word: DictionaryWord) {
        words.removeAll { $0.id == word.id }
        save()

        guard apiClient != nil else { return }
        tombstones.insert(word.id)
        saveTombstones()
        Task { await self.deleteWordOnServer(word.id) }
    }

    func clearLocalCache() {
        words.removeAll()
        tombstones.removeAll()
        save()
        saveTombstones()
    }

    /// Pull server state, merge, push local-only words, retry tombstoned deletes.
    func hydrate() async {
        guard let apiClient else { return }

        let serverWords: [DictionaryWordDTO]
        do {
            serverWords = try await apiClient.listDictationDictionary()
        } catch {
            log.error("Hydrate dictionary fetch failed: \(error.localizedDescription)")
            return
        }

        for serverWord in serverWords where tombstones.contains(serverWord.clientWordID) {
            do {
                try await apiClient.deleteDictionaryWord(clientWordID: serverWord.clientWordID)
                tombstones.remove(serverWord.clientWordID)
            } catch {
                log.error("Hydrate dictionary tombstone replay failed: \(error.localizedDescription)")
            }
        }
        saveTombstones()

        let localIDs = Set(words.map(\.id))
        let serverByID = Dictionary(uniqueKeysWithValues: serverWords.map { ($0.clientWordID, $0) })
        var additions: [DictionaryWord] = []
        for (id, dto) in serverByID where !localIDs.contains(id) && !tombstones.contains(id) {
            additions.append(
                DictionaryWord(
                    id: dto.clientWordID,
                    word: dto.word,
                    replacement: dto.replacement,
                    createdAt: dto.occurredAt
                )
            )
        }
        if !additions.isEmpty {
            words.append(contentsOf: additions)
            words.sort { $0.word.localizedCaseInsensitiveCompare($1.word) == .orderedAscending }
            save()
        }

        let serverIDs = Set(serverByID.keys)
        for entry in words where !serverIDs.contains(entry.id) {
            await pushWord(entry)
        }
    }

    /// Vocabulary list for prompt conditioning — sent to transcription provider
    var vocabularyList: [String] {
        words.map(\.word)
    }

    /// Apply replacement rules to transcribed text.
    ///
    /// Matches whole words only. A literal substring replace would rewrite
    /// inside words ("cat" → "dog" turning "category" into "dogegory"), so we
    /// anchor on Unicode word boundaries — `.useUnicodeWordBoundaries` is
    /// essential or `\b` is ASCII-only and never fires between Cyrillic letters.
    func applyReplacements(to text: String) -> String {
        var result = text
        for word in words where word.isReplacement {
            guard let replacement = word.replacement,
                  !word.word.isEmpty else { continue }
            let pattern = "\\b" + NSRegularExpression.escapedPattern(for: word.word) + "\\b"
            guard let regex = try? NSRegularExpression(
                pattern: pattern,
                options: [.caseInsensitive, .useUnicodeWordBoundaries]
            ) else { continue }
            let range = NSRange(result.startIndex..., in: result)
            result = regex.stringByReplacingMatches(
                in: result,
                range: range,
                withTemplate: NSRegularExpression.escapedTemplate(for: replacement)
            )
        }
        return result
    }

    // MARK: - Sync helpers

    private static let iso8601: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private func pushWord(_ word: DictionaryWord) async {
        guard let apiClient else { return }
        let request = CreateDictionaryWordRequest(
            clientWordID: word.id,
            word: word.word,
            replacement: word.replacement,
            occurredAt: Self.iso8601.string(from: word.createdAt)
        )
        do {
            _ = try await apiClient.createDictionaryWord(request)
        } catch {
            log.error("Push word failed (will retry on next hydrate): \(error.localizedDescription)")
        }
    }

    private func deleteWordOnServer(_ id: UUID) async {
        guard let apiClient else { return }
        do {
            try await apiClient.deleteDictionaryWord(clientWordID: id)
            tombstones.remove(id)
            saveTombstones()
        } catch {
            log.error("Delete word on server failed (tombstone remains): \(error.localizedDescription)")
        }
    }

    // MARK: - Persistence

    private func load() {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return }
        do {
            let data = try Data(contentsOf: fileURL)
            words = try JSONDecoder().decode([DictionaryWord].self, from: data)
            log.info("Loaded \(self.words.count) dictionary words")
        } catch {
            log.error("Failed to load dictionary: \(error.localizedDescription)")
        }
    }

    private func save() {
        do {
            try FileManager.default.createDirectory(
                at: fileURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            let data = try JSONEncoder().encode(words)
            try data.write(to: fileURL, options: .atomic)
        } catch {
            log.error("Failed to save dictionary: \(error.localizedDescription)")
        }
    }

    private func loadTombstones() {
        guard FileManager.default.fileExists(atPath: tombstonesURL.path) else { return }
        do {
            let data = try Data(contentsOf: tombstonesURL)
            tombstones = Set(try JSONDecoder().decode([UUID].self, from: data))
        } catch {
            log.error("Failed to load dictionary tombstones: \(error.localizedDescription)")
        }
    }

    private func saveTombstones() {
        do {
            try FileManager.default.createDirectory(
                at: tombstonesURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            let data = try JSONEncoder().encode(Array(tombstones))
            try data.write(to: tombstonesURL, options: .atomic)
        } catch {
            log.error("Failed to save dictionary tombstones: \(error.localizedDescription)")
        }
    }
}
